File: packages/provide-uterm/src/provide/uterm/channels.py
#
# SPDX-FileCopyrightText: Copyright (c) 2025-2026 provide.io llc. All rights reserved.
# SPDX-License-Identifier: AGPL-3.0-or-later
#
"""Generic typed-channel negotiation over the inline control channel."""

from __future__ import annotations

from collections.abc import Mapping
from dataclasses import dataclass
from typing import Any

from provide.uterm.control_channel import ControlChunk, ControlFrameDecoder, is_control_frame


@dataclass(frozen=True, slots=True)
class ChannelHello:
    """Client-advertised typed-channel versions."""

    channels: dict[str, int]


class NegotiatedChannels:
    """Per-connection typed-channel grants and sequence counters."""

    def __init__(self, supported: Mapping[str, int], *, default_channel: str | None = None) -> None:
        self._supported = _normalize_supported(supported)
        if default_channel is not None and default_channel not in self._supported:
            raise ValueError(f"default channel is not supported: {default_channel!r}")
        self._default_channel = default_channel
        self._granted: dict[str, int] = {}
        self._seq: dict[str, int] = {}

    @property
    def granted(self) -> dict[str, int]:
        """Return a copy of the currently granted channels."""
        return dict(self._granted)

    def is_negotiated(self, channel: str | None = None) -> bool:
        """Return whether *channel* is negotiated, defaulting to the configured primary channel."""
        selected = self._select_channel(channel)
        return selected in self._granted

    def handle_hello(
        self,
        hello: ChannelHello | Mapping[str, Any],
        *,
        ack_fields: Mapping[str, Any] | None = None,
    ) -> dict[str, Any]:
        """Negotiate channel versions and return a ``hello_ack`` payload."""
        extra = dict(ack_fields or {})
        reserved = {"type", "channels"} & extra.keys()
        if reserved:
            raise ValueError(f"reserved hello_ack field: {sorted(reserved)[0]}")
        requested = hello.channels if isinstance(hello, ChannelHello) else _coerce_channel_map(hello.get("channels"))
        self._granted = _negotiate(self._supported, requested)
        return {"type": "hello_ack", "channels": dict(self._granted), **extra}

    def next_seq(self, channel: str | None = None) -> int:
        """Increment and return the sequence number for *channel*."""
        selected = self._select_channel(channel)
        self._seq[selected] = self._seq.get(selected, 0) + 1
        return self._seq[selected]

    def export_grants(self) -> dict[str, int]:
        """Return a serializable granted-channel map."""
        return dict(self._granted)

    def restore_grants(self, grants: Mapping[str, int]) -> None:
        """Restore persisted grants and reset sequence counters for a fresh channel instance."""
        self._granted = _negotiate(self._supported, _coerce_channel_map(grants))
        self._seq = {}

    def _select_channel(self, channel: str | None) -> str:
        selected = channel if channel is not None else self._default_channel
        if selected is None:
            raise ValueError("channel is required when no default_channel is configured")
        return selected


def parse_channel_hello(raw: str) -> ChannelHello | None:
    """Parse a framed ``hello`` payload, or return ``None`` when it is not a channel hello."""
    if not raw or not is_control_frame(raw):
        return None
    decoder = ControlFrameDecoder()
    try:
        chunks = decoder.feed(raw)
        chunks += decoder.finish()
    except Exception:
        return None
    for chunk in chunks:
        if isinstance(chunk, ControlChunk) and chunk.control.get("type") == "hello":
            try:
                return ChannelHello(channels=_coerce_channel_map(chunk.control.get("channels")))
            except ValueError:
                return None
    return None


def _normalize_supported(supported: Mapping[str, int]) -> dict[str, int]:
    normalized = _coerce_channel_map(supported)
    if not normalized:
        raise ValueError("at least one supported channel is required")
    return normalized


def _coerce_channel_map(value: Any) -> dict[str, int]:
    if not isinstance(value, Mapping):
        raise ValueError("channels must be a mapping")
    channels: dict[str, int] = {}
    for name, version in value.items():
        if not isinstance(name, str) or not name:
            raise ValueError("channel names must be non-empty strings")
        if isinstance(version, bool) or not isinstance(version, int):
            raise ValueError("channel versions must be integers")
        channels[name] = version
    return channels


def _negotiate(supported: Mapping[str, int], requested: Mapping[str, int]) -> dict[str, int]:
    granted: dict[str, int] = {}
    for name, version in requested.items():
        supported_version = supported.get(name)
        if supported_version is not None and version > 0:
            granted[name] = min(version, supported_version)
    return granted

File: packages/provide-uterm-ts/src/channels/channels.ts
//
// SPDX-FileCopyrightText: Copyright (c) 2025-2026 provide.io llc. All rights reserved.
// SPDX-License-Identifier: AGPL-3.0-or-later
//

/**
 * Generic typed-channel negotiation over the inline control channel.
 *
 * Port of the Python module `provide.uterm.channels` and the Go package
 * `channels`.
 */

import type { ControlFrameChunk } from "../control-channel/index.ts";
import { ControlFrameDecoder, isControlFrame } from "../control-channel/index.ts";

/** Client-advertised typed-channel versions. */
export interface ChannelHello {
  readonly channels: Record<string, number>;
}

/** Fields the `hello_ack` payload owns and callers may not supply. */
const RESERVED_ACK_FIELDS = ["channels", "type"] as const;

/** Construction options for {@link NegotiatedChannels}. */
export interface NegotiatedChannelsOptions {
  /** Channel used when a call does not name one. */
  defaultChannel?: string;
}

/**
 * Coerce a channel map, rejecting anything the reference would reject.
 *
 * JavaScript needs one guard CPython does not: an array is an object here,
 * so it has to be refused explicitly to match `isinstance(value, Mapping)`.
 * Booleans are refused for the same reason CPython refuses them — `True` is
 * an `int` in Python, and a boolean version is a caller mistake either way.
 */
function coerceChannelMap(value: unknown): Record<string, number> {
  if (typeof value !== "object" || value === null || Array.isArray(value)) {
    throw new Error("channels must be a mapping");
  }
  const channels: Record<string, number> = {};
  for (const [name, version] of Object.entries(value)) {
    if (name === "") {
      throw new Error("channel names must be non-empty strings");
    }
    if (typeof version !== "number" || !Number.isInteger(version)) {
      throw new Error("channel versions must be integers");
    }
    channels[name] = version;
  }
  return channels;
}

/** Grant the lower of each requested and supported version, dropping the rest. */
function negotiate(
  supported: Readonly<Record<string, number>>,
  requested: Readonly<Record<string, number>>,
): Record<string, number> {
  const granted: Record<string, number> = {};
  for (const [name, version] of Object.entries(requested)) {
    const supportedVersion = supported[name];
    if (supportedVersion !== undefined && version > 0) {
      granted[name] = Math.min(version, supportedVersion);
    }
  }
  return granted;
}

/** Per-connection typed-channel grants and sequence counters. */
export class NegotiatedChannels {
  readonly #supported: Record<string, number>;
  readonly #defaultChannel: string | undefined;
  #granted: Record<string, number> = {};
  #seq = new Map<string, number>();

  constructor(supported: Record<string, number>, options: NegotiatedChannelsOptions = {}) {
    this.#supported = coerceChannelMap(supported);
    if (Object.keys(this.#supported).length === 0) {
      throw new Error("at least one supported channel is required");
    }
    const defaultChannel = options.defaultChannel;
    if (defaultChannel !== undefined && !(defaultChannel in this.#supported)) {
      throw new Error(`default channel is not supported: ${JSON.stringify(defaultChannel)}`);
    }
    this.#defaultChannel = defaultChannel;
  }

  /** A copy of the currently granted channels. */
  get granted(): Record<string, number> {
    return { ...this.#granted };
  }

  /** Resolve `channel`, falling back to the configured default. */
  #selectChannel(channel: string | undefined): string {
    const selected = channel ?? this.#defaultChannel;
    if (selected === undefined) {
      throw new Error("channel is required when no default_channel is configured");
    }
    return selected;
  }

  /** Whether `channel` is negotiated, defaulting to the primary channel. */
  isNegotiated(channel?: string): boolean {
    return this.#selectChannel(channel) in this.#granted;
  }

  /** Negotiate channel versions and build the `hello_ack` payload. */
  handleHello(hello: unknown, ackFields: Record<string, unknown> = {}): Record<string, unknown> {
    const reserved = RESERVED_ACK_FIELDS.filter((field) => field in ackFields);
    if (reserved.length > 0) {
      throw new Error(`reserved hello_ack field: ${reserved[0]}`);
    }
    const payload = hello as { channels?: unknown };
    this.#granted = negotiate(this.#supported, coerceChannelMap(payload.channels));
    return { type: "hello_ack", channels: { ...this.#granted }, ...ackFields };
  }

  /** Increment and return the sequence number for `channel`. */
  nextSeq(channel?: string): number {
    const selected = this.#selectChannel(channel);
    const next = (this.#seq.get(selected) ?? 0) + 1;
    this.#seq.set(selected, next);
    return next;
  }

  /** A serialisable granted-channel map. */
  exportGrants(): Record<string, number> {
    return { ...this.#granted };
  }

  /**
   * Restore persisted grants, re-negotiated against what this instance
   * supports, and reset the sequence counters for the fresh connection.
   */
  restoreGrants(grants: Record<string, number>): void {
    this.#granted = negotiate(this.#supported, coerceChannelMap(grants));
    this.#seq = new Map();
  }
}

/**
 * Parse a framed `hello` payload, or return `null` when it is not a channel
 * hello or cannot be read as one.
 */
export function parseChannelHello(raw: string): ChannelHello | null {
  if (raw === "" || !isControlFrame(raw)) {
    return null;
  }
  const decoder = new ControlFrameDecoder();
  let chunks: ControlFrameChunk[];
  try {
    chunks = [...decoder.feed(raw), ...decoder.finish()];
  } catch {
    return null;
  }
  for (const chunk of chunks) {
    if (chunk.kind === "control" && chunk.control.type === "hello") {
      try {
        return { channels: coerceChannelMap(chunk.control.channels) };
      } catch {
        return null;
      }
    }
  }
  return null;
}
